src/ray/object_manager/object_directory.cc

```diff
@@ -8,30 +8,30 @@ ObjectDirectory::ObjectDirectory(boost::asio::io_service &io_service,
 
 namespace {
 
-/// Process a suffix of the object table log and store the result in
+/// Process a notification of the object table entries and store the result in
 /// client_ids. This assumes that client_ids already contains the result of the
-/// object table log up to but not including this suffix. This also stores a
+/// object table entries up to but not including this notification. This also stores a
 /// bool in has_been_created indicating whether the object has ever been
 /// created before.
-void UpdateObjectLocations(const std::vector<ObjectTableDataT> &location_history,
+void UpdateObjectLocations(const GcsTableNotificationMode mode,
+                           const std::vector<ObjectTableDataT> &location_updates,
                            const ray::gcs::ClientTable &client_table,
                            std::unordered_set<ClientID> *client_ids,
                            bool *has_been_created) {
-  // location_history contains the history of locations of the object (it is a log),
-  // which might look like the following:
-  //   client1.is_eviction = false
-  //   client1.is_eviction = true
-  //   client2.is_eviction = false
-  // In such a scenario, we want to indicate client2 is the only client that contains
-  // the object, which the following code achieves.
-  if (!location_history.empty()) {
+  // location_updates contains the updates of locations of the object.
+  // with GcsTableNotificationMode, we can determine whether the update mode is
+  // addition or deletion (or just current value).
+  if (!location_updates.empty()) {
     // If there are entries, then the object has been created. Once this flag
     // is set to true, it should never go back to false.
     *has_been_created = true;
   }
-  for (const auto &object_table_data : location_history) {
+  if (mode == GcsTableNotificationMode::CURRENT_VALUE) {
```

Comment:
Can you add a comment explaining why we should clear `client_ids`? This wasn't immediately obvious to me.

Reply:
If the caller invokes `UpdateObjectLocations` with `CURRENT_VALUE` and a non-empty `client_ids`, and `client_ids` has more elements than `location_updates`, this means some clients were removed (during redis reconnection process). Anyway, I'll remove `CURRENT_VALUE` for simplicity.